File: server/privacy.ts
/**
 * Agent-private note blocks.
 *
 * Human-authenticated clients receive the stored Markdown unchanged. Agent
 * access receives stable placeholders so an agent can edit around a private
 * block without seeing or accidentally erasing it.
 *
 * Syntax:
 *   :::private
 *   secret material
 *   :::
 */

export type PrivateBlock = {
  from: number;
  to: number;
  raw: string;
  id: string;
};

const START = /^[\t ]*:::private[\t ]*$/i;
const END = /^[\t ]*:::[\t ]*$/;
const PLACEHOLDER_ID = /\[Private block hidden from agents\. id=([a-z0-9-]+)\]/gi;

function stableBlockId(raw: string, index: number): string {
  // This is an edit-preservation handle, not a secret hash. Avoid a crypto
  // dependency so the parser stays deterministic in every server test/runtime.
  let hash = 2166136261;
  const input = `${index}\0${raw}`;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return `p${(hash >>> 0).toString(36)}-${index + 1}`;
}

type Line = { from: number; to: number; text: string };

function contentLines(content: string): Line[] {
  if (!content) return [];
  const lines: Line[] = [];
  let from = 0;
  while (from < content.length) {
    const newline = content.indexOf('\n', from);
    const to = newline === -1 ? content.length : newline;
    const raw = content.slice(from, to);
    lines.push({ from, to, text: raw.endsWith('\r') ? raw.slice(0, -1) : raw });
    if (newline === -1) break;
    from = newline + 1;
  }
  return lines;
}

export function privateBlocks(content: string): PrivateBlock[] {
  const lines = contentLines(String(content || ''));
  const blocks: PrivateBlock[] = [];
  for (let i = 0; i < lines.length; i += 1) {
    if (!START.test(lines[i].text)) continue;
    let end = content.length;
    let closingLine = lines.length;
    for (let j = i + 1; j < lines.length; j += 1) {
      if (!END.test(lines[j].text)) continue;
      end = lines[j].to;
      closingLine = j;
      break;
    }
    const raw = content.slice(lines[i].from, end);
    blocks.push({
      from: lines[i].from,
      to: end,
      raw,
      id: stableBlockId(raw, blocks.length),
    });
    // An unterminated private block is fail-closed: everything after it is
    // private, so there cannot be another independently parsed block.
    if (closingLine >= lines.length) break;
    i = closingLine;
  }
  return blocks;
}

function agentPlaceholder(block: PrivateBlock): string {
  return [
    ':::private',
    `[Private block hidden from agents. id=${block.id}]`,
    ':::',
  ].join('\n');
}

function isAgentPlaceholder(raw: string): boolean {
  PLACEHOLDER_ID.lastIndex = 0;
  return PLACEHOLDER_ID.test(raw);
}

function replaceBlocks(
  content: string,
  replacement: (block: PrivateBlock) => string,
): string {
  const blocks = privateBlocks(content);
  if (!blocks.length) return content;
  let cursor = 0;
  let result = '';
  for (const block of blocks) {
    result += content.slice(cursor, block.from);
    result += replacement(block);
    cursor = block.to;
  }
  return result + content.slice(cursor);
}

export function redactPrivateBlocks(content: string): string {
  return replaceBlocks(
    String(content || ''),
    (block) => isAgentPlaceholder(block.raw) ? block.raw : agentPlaceholder(block),
  );
}

/**
 * `content_preview` is whitespace-collapsed in the database, so line-oriented
 * closing markers are no longer trustworthy. Once an opener appears, redact
 * the remainder of the preview (fail closed).
 */
export function redactPrivatePreview(content: string): string {
  const value = String(content || '');
  const opener = value.search(/:::private\b/i);
  if (opener < 0) return value;
  return `${value.slice(0, opener)}${'[Private block hidden from agents]'}`.trim();
}

export function redactPrivateBlocksForPublic(content: string): string {
  return replaceBlocks(
    String(content || ''),
    () => '> Private block omitted from the public note.',
  );
}

/**
 * Rehydrate placeholders in an agent-authored full-note update.
 *
 * Existing blocks are mandatory and single-use. Removing, duplicating, or
 * editing a placeholder fails instead of silently deleting the secret.
 */
export function restoreAgentPrivateBlocks(existing: string, incoming: string): string {
  const blocks = privateBlocks(existing);
  if (!blocks.length) {
    PLACEHOLDER_ID.lastIndex = 0;
    if (PLACEHOLDER_ID.test(incoming)) {
      throw new Error('Unknown private block placeholder.');
    }
    return incoming;
  }

  const expectedIds = new Set(blocks.map((block) => block.id));
  PLACEHOLDER_ID.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = PLACEHOLDER_ID.exec(incoming)) !== null) {
    if (!expectedIds.has(match[1])) throw new Error('Unknown private block placeholder.');
  }

  let restored = incoming;
  for (const block of blocks) {
    const placeholder = agentPlaceholder(block);
    const occurrences = restored.split(placeholder).length - 1;
    if (occurrences !== 1) {
      throw new Error('Agent edits must preserve every private block placeholder exactly once.');
    }
    restored = restored.replace(placeholder, block.raw);
  }
  return restored;
}

/** Incoming-first alias used by the HTTP note update route. */
export function restorePrivateBlocks(incoming: string, existing: string): string {
  return restoreAgentPrivateBlocks(existing, incoming);
}

/** Recursively redact privacy blocks in any JSON response sent to an agent. */
export function sanitizeAgentJson<T>(value: T): T {
  if (typeof value === 'string') return redactPrivateBlocks(value) as T;
  if (Array.isArray(value)) return value.map((item) => sanitizeAgentJson(item)) as T;
  if (value && typeof value === 'object') {
    const output: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value)) {
      output[key] = key === 'content_preview' && typeof item === 'string'
        ? redactPrivatePreview(item)
        : sanitizeAgentJson(item);
    }
    return output as T;
  }
  return value;
}

/**
 * Restricted JWTs are capabilities for the three agent helpers, not general
 * user sessions. Keep this list intentionally narrow.
 */
export function isAgentApiRequestAllowed(methodRaw: string, path: string): boolean {
  const method = methodRaw.toUpperCase();
  const rules: Array<[string[], RegExp]> = [
    [['GET'], /^\/api\/vaults$/],
    [['GET'], /^\/api\/vaults\/[^/]+$/],
    [['GET'], /^\/api\/vaults\/[^/]+\/(?:folders|notes|search|tags)$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/folders$/],
    // Scoped folder mutation for the current agent's writable vault. The
    // handler still enforces vault membership and rejects viewer roles.
    [['PATCH'], /^\/api\/folders\/[^/]+$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/notes$/],
    [['GET', 'PUT'], /^\/api\/vaults\/[^/]+\/agent-memory$/],
    [['GET', 'POST'], /^\/api\/vaults\/[^/]+\/scratchpad(?:\/[^/]+)*(?:\/close)?$/],
    [['GET', 'POST', 'PATCH'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/messages(?:\/[^/]+)?$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/messages\/[^/]+\/collaborate$/],
    [['GET'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/agents$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions$/],
    [['GET'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions$/],
    [['GET'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions\/[^/]+$/],
    [['GET'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions\/[^/]+\/history$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions\/[^/]+\/(?:tasks|finish)$/],
    [['PATCH'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/missions\/tasks\/[^/]+$/],
    [['POST'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/distill$/],
    [['PUT'], /^\/api\/vaults\/[^/]+\/channels\/[^/]+\/agents\/[^/]+\/avatar$/],
    [['GET', 'PUT', 'DELETE'], /^\/api\/notes\/[^/]+$/],
    [['POST'], /^\/api\/notes\/[^/]+\/(?:rename|move|unlist|pin|archive)$/],
    [['POST', 'DELETE'], /^\/api\/notes\/[^/]+\/tags(?:\/[^/]+)?$/],
    [['GET'], /^\/api\/notes\/[^/]+\/backlinks$/],
    [['GET'], /^\/api\/notes\/[^/]+\/assets\/[^/]+$/],
  ];
  return rules.some(([methods, pattern]) => methods.includes(method) && pattern.test(path));
}

File: backend_elixir/lib/cascade/content/privacy.ex
defmodule Cascade.Content.Privacy do
  @moduledoc false

  import Bitwise

  @start ~r/^[\t ]*:::private[\t ]*$/iu
  @finish ~r/^[\t ]*:::[\t ]*$/u
  @placeholder ~r/\[Private block hidden from agents\. id=([a-z0-9-]+)\]/iu

  def redact_note(nil, _agent?), do: nil

  def redact_note(note, agent?) do
    note = Map.delete(note, :file_path)

    if agent? do
      note
      |> maybe_update(:content, &redact_blocks/1)
      |> maybe_update(:content_preview, &redact_preview/1)
    else
      note
    end
  end

  def redact_blocks(content) do
    replace_blocks(to_string(content), fn block ->
      if Regex.match?(@placeholder, block.raw), do: block.raw, else: placeholder(block)
    end)
  end

  def redact_preview(content) do
    value = to_string(content)

    case :binary.match(String.downcase(value), ":::private") do
      :nomatch ->
        value

      {offset, _length} ->
        String.trim(binary_part(value, 0, offset) <> "[Private block hidden from agents]")
    end
  end

  def sanitize_json(value) when is_binary(value), do: redact_blocks(value)
  def sanitize_json(value) when is_list(value), do: Enum.map(value, &sanitize_json/1)

  def sanitize_json(value) when is_map(value) do
    Map.new(value, fn
      {key, item} when key in [:content_preview, "content_preview"] and is_binary(item) ->
        {key, redact_preview(item)}

      {key, item} ->
        {key, sanitize_json(item)}
    end)
  end

  def sanitize_json(value), do: value

  def restore_blocks(existing, incoming) do
    blocks = private_blocks(existing)
    incoming = to_string(incoming)

    if blocks == [] do
      if Regex.match?(@placeholder, incoming),
        do: raise(ArgumentError, "Unknown private block placeholder.")

      incoming
    else
      expected = blocks |> Enum.map(& &1.id) |> MapSet.new()

      @placeholder
      |> Regex.scan(incoming, capture: :all_but_first)
      |> Enum.each(fn [id] ->
        if not MapSet.member?(expected, id),
          do: raise(ArgumentError, "Unknown private block placeholder.")
      end)

      Enum.reduce(blocks, incoming, fn block, restored ->
        marker = placeholder(block)

        if length(:binary.matches(restored, marker)) != 1 do
          raise ArgumentError,
                "Agent edits must preserve every private block placeholder exactly once."
        end

        String.replace(restored, marker, block.raw, global: false)
      end)
    end
  end

  defp private_blocks(content) do
    lines = content_lines(to_string(content))
    parse_blocks(lines, to_string(content), 0, [])
  end

  defp parse_blocks(lines, _content, index, blocks) when index >= length(lines),
    do: Enum.reverse(blocks)

  defp parse_blocks(lines, content, index, blocks) do
    line = Enum.at(lines, index)

    if Regex.match?(@start, line.text) do
      {finish, closing_index} = find_finish(lines, index + 1, byte_size(content))
      raw = binary_part(content, line.from, finish - line.from)
      block = %{from: line.from, to: finish, raw: raw, id: stable_id(raw, length(blocks))}

      if closing_index >= length(lines) do
        Enum.reverse([block | blocks])
      else
        parse_blocks(lines, content, closing_index + 1, [block | blocks])
      end
    else
      parse_blocks(lines, content, index + 1, blocks)
    end
  end

  defp find_finish(lines, index, default) when index >= length(lines),
    do: {default, length(lines)}

  defp find_finish(lines, index, default) do
    line = Enum.at(lines, index)

    if Regex.match?(@finish, line.text),
      do: {line.to, index},
      else: find_finish(lines, index + 1, default)
  end

  defp content_lines(""), do: []

  defp content_lines(content) do
    do_content_lines(content, 0, [])
  end

  defp do_content_lines(content, from, lines) when from >= byte_size(content),
    do: Enum.reverse(lines)

  defp do_content_lines(content, from, lines) do
    rest = binary_part(content, from, byte_size(content) - from)

    case :binary.match(rest, "\n") do
      :nomatch ->
        raw = rest

        text =
          if String.ends_with?(raw, "\r"), do: binary_part(raw, 0, byte_size(raw) - 1), else: raw

        Enum.reverse([%{from: from, to: byte_size(content), text: text} | lines])

      {relative, 1} ->
        raw = binary_part(content, from, relative)

        text =
          if String.ends_with?(raw, "\r"), do: binary_part(raw, 0, byte_size(raw) - 1), else: raw

        do_content_lines(content, from + relative + 1, [
          %{from: from, to: from + relative, text: text} | lines
        ])
    end
  end

  defp stable_id(raw, index) do
    input = Integer.to_string(index) <> <<0>> <> raw

    utf16 = :unicode.characters_to_binary(input, :utf8, {:utf16, :little})

    hash =
      for <<unit::little-16 <- utf16>>, reduce: 2_166_136_261 do
        current -> band(bxor(current, unit) * 16_777_619, 0xFFFFFFFF)
      end

    "p#{Integer.to_string(hash, 36)}-#{index + 1}"
  end

  defp placeholder(block) do
    ":::private\n[Private block hidden from agents. id=#{block.id}]\n:::"
  end

  defp replace_blocks(content, replacement) do
    blocks = private_blocks(content)

    {parts, cursor} =
      Enum.reduce(blocks, {[], 0}, fn block, {parts, cursor} ->
        prefix = binary_part(content, cursor, block.from - cursor)
        {[parts, prefix, replacement.(block)], block.to}
      end)

    IO.iodata_to_binary([parts, binary_part(content, cursor, byte_size(content) - cursor)])
  end

  defp maybe_update(map, key, function) do
    case Map.fetch(map, key) do
      {:ok, value} when is_binary(value) -> Map.put(map, key, function.(value))
      _ -> map
    end
  end
end
